src/modules/bulletproofs/tests_impl.h

```diff
@@ -0,0 +1,545 @@
+    cs = (secp256k1_scalar*)secp256k1_scratch_alloc(&ctx->error_callback, scratch, h_len * sizeof(secp256k1_scalar));
+    gs = (secp256k1_ge*)secp256k1_scratch_alloc(&ctx->error_callback, scratch, (g_len + h_len) * sizeof(secp256k1_ge));
+    if (ns == NULL || ls == NULL || cs == NULL || gs == NULL) {
+        secp256k1_scratch_apply_checkpoint(&ctx->error_callback, scratch, scratch_checkpoint);
+        return 0;
+    }
+    memcpy(ns, n_vec, g_len * sizeof(secp256k1_scalar));
+    memcpy(ls, l_vec, h_len * sizeof(secp256k1_scalar));
+    memcpy(cs, c_vec, h_len * sizeof(secp256k1_scalar));
+    memcpy(gs, gens_vec->gens, (g_len + h_len) * sizeof(secp256k1_ge));
+
+    /* Commit to the initial public values */
+    secp256k1_norm_arg_commit_initial_data(&transcript, &r, gens_vec, g_len, c_vec, c_vec_len, &comm);
+
+    res = secp256k1_bppp_rangeproof_norm_product_prove(
+        ctx,
+        scratch,
+        proof,
+        proof_len,
+        &transcript, /* Transcript hash of the parent protocol */
+        &r,
+        gs,
+        gens_vec->n,
+        ns,
+        n_vec_len,
+        ls,
+        l_vec_len,
+        cs,
+        c_vec_len
+    );
+    secp256k1_scratch_apply_checkpoint(&ctx->error_callback, scratch, scratch_checkpoint);
+    return res;
+}
+
+/* Verify the proof */
+static int secp256k1_norm_arg_verify(
+    secp256k1_scratch_space* scratch,
+    const unsigned char* proof,
+    size_t proof_len,
+    const secp256k1_scalar* r_ch,
+    const secp256k1_bulletproofs_generators* gens_vec,
+    size_t g_len,
+    const secp256k1_scalar* c_vec,
+    size_t c_vec_len,
+    const secp256k1_ge* commit
+) {
+    secp256k1_ge comm = *commit;
+    int res;
+    secp256k1_sha256 transcript;
+    secp256k1_scalar r = *r_ch;
+
+    /* Commit to the initial public values */
+    secp256k1_norm_arg_commit_initial_data(&transcript, &r, gens_vec, g_len, c_vec, c_vec_len, &comm);
+
+    res = secp256k1_bppp_rangeproof_norm_product_verify(
+        ctx,
+        scratch,
+        proof,
+        proof_len,
+        &transcript,
+        &r,
+        gens_vec,
+        g_len,
+        c_vec,
+        c_vec_len,
+        commit
+    );
+    return res;
+}
+
+void norm_arg_zero(void) {
+    secp256k1_scalar n_vec[64], l_vec[64], c_vec[64];
+    secp256k1_scalar r, q;
+    secp256k1_ge commit;
+    size_t i;
+    secp256k1_scratch *scratch = secp256k1_scratch_space_create(ctx, 1000*10); /* shouldn't need much */
+    unsigned char proof[1000];
+
+    random_scalar_order(&r);
+    secp256k1_scalar_sqr(&q, &r);
+
+    /* l and n are the zero vector */
+    {
+        unsigned int n_vec_len = 1;
+        unsigned int c_vec_len = 1;
+        secp256k1_bulletproofs_generators *gs = secp256k1_bulletproofs_generators_create(ctx, n_vec_len + c_vec_len);
+        size_t plen = sizeof(proof);
+        for (i = 0; i < n_vec_len; i++) {
+            secp256k1_scalar_set_int(&n_vec[i], 1);
```

Comment:
Is there a reason you changed this from 0 to 1? The test description says that `n_vec[i]` should be 0.

Reply:
Yes, because the bulletproofs commitment was the point at infinity, because all `n`, `l` and `v` are zero. The new serialization code as per @real-or-random 's suggestion `ec_pubkey_serialize` correctly fails to serialize this. I had to update to test to reflect this. 

My understanding was this test should only check that this test only produces two scalars which are input `n[0]` and `l[0]` directly. Them being zero really does not change anything.

Reply:
I thought I added in the commit message description, but seems like not :( .

Reply:
The intention of the test was actually to test what happens if you pass zero vectors. The new test is somewhat redundant as we already have a static test vector. So I think it would be preferable to reflect the `ec_pubkey_serialize` change in the tests by checking that the functions fail as expected.

Reply:
I added a commit to https://github.com/jonasnick/secp256k1-zkp/commits/b%2B%2B-jn-3 that allows the commitment to be the point at infinity again. The reason is that we want the norm argument to be usable by a variety of protocols which don't necessarily forbid serializing the point at infinity into the transcript. In particular, for verify and prove we shouldn't rely on the outer protocol to serialize points in a certain way. Therefore, it's better to test verify and prove for the zero vector (good to test the edges anyway) which required that `commit()` succeeds even if its output is the point at infinity.